### DISCUSSION TITLE
SERIAL TEST:   osnma-longan-nano-client > no   .....  CuteCom app > YES!

### DISCUSSION THREAD

**Discussion by K4KDR** [daniestevez/galileo-osnma#16] [Q&A]
With the longan-nano connected to a proper 3.3v USB/UART adapter, it presented correctly to my computer as /dev/ttyUSB0.

However, upon running the osnma-longan-nano-client utility, there was no READY reply when I reset the nano and attempts to send your test strings to the device via serial resulted in:

```
k4kdr@3010i5:~/galileo-osnma/osnma-longan-nano-client/target/debug$ ./osnma-longan-nano-client /dev/ttyUSB0
19 1176 120939 2a2aaaaaaaaaaaaaaaaaa80327fd4618
Error: Custom { kind: InvalidData, error: "incorrect galmon magic value" }
```

... the result was the same whether the client app was built with the default  **cargo build** or **cargo build --release**

However, if a standard serial comm app is used such as **CuteCom**, the code running on the nano works perfectly!!

![cutecom-nano-both-test-strings](https://user-images.githubusercontent.com/19886368/161395536-ff3fb7cf-9dc5-4ce3-a65c-57e6f0ddc30d.png)

... so I was _extremely_ pleased about that.

Any suggestions on what I should check to determine why the **osnma-longan-nano-client app** is not establishing a connection to the nano for me?

Many thanks!

**Comment by daniestevez** [comment]
You're not running `osnma-longan-nano-client` correctly. Rather than giving it  `19 1176 120939 2a2aaaaaaaaaaaaaaaaaa80327fd4618` as parameters (or typing this), you should give it data in the Galmon protocol, in the same way you would do for `galmon-osnma`. So you can either
```
nc 86.82.68.237 10000 | ./osnma-longan-nano-client /dev/ttyUSB0
```
or
```
<data.pb ./osnma-longan-nano-client /dev/ttyUSB0
```
using [this sample file](http://eala.destevez.net/~daniel/ubxtool-inav.20220307.pb) or another one you have as `data.pb`.

**Comment by daniestevez** [comment]
Take a look at [this part of the README](https://github.com/daniestevez/galileo-osnma/tree/main/osnma-longan-nano#running-the-serial-port-client), since maybe you have an idea about how it could be written more clearly.

**Comment by K4KDR** [comment]
Thank you!  Oh my, this longan-nano is extremely fast.  Using the 'nc' method to leverage the web data stream, the **AUTH** messages just fly by!

![nano-running-from-nc-web-stream](https://user-images.githubusercontent.com/19886368/161403433-8dcf7944-22f3-40d4-8ccc-fb7e4812a092.png)

... and yes, I will be glad to make suggestions ref. that README.   I have forked the repository and will attempt to contribute once it is clear that I have done everything correct across the various test scenarios!

**Comment by daniestevez** [comment]
It depends on what you consider fast hahaha. I have a file with more than a week worth of data, and the longan-nano takes hours (maybe > 1 day) to go through that, while my laptop processes it in a few seconds. I think that the bottleneck is perhaps the UART. I'm curious how much time the actual computations take.

**Comment by daniestevez** [comment]
I'm trying to clean up the issues by closing what is already solved and moving to a discussion what has background information that could be interesting to others. I think this issue can be closed, since already we added more information to the README.

**Comment by K4KDR** [comment]
Agree!

Github doesn't have a button for the non-owner of the repository to move an ISSUE to a DISCUSSION, but please feel free.
